end_to_end/test_llama2_7b.sh

```diff
@@ -0,0 +1,69 @@
+
+# We also run pre-training of Llama2-7b, this is similar to the finetuning command except we don't pass any checkpoint directory to load parameters from
+python3 MaxText/train.py MaxText/configs/base.yml run_name=runner_pretraining_${idx}  per_device_batch_size=1 model_name='llama2-7b' ici_tensor_parallelism=4 steps=10 max_target_length=1024 per_device_batch_size=1
+
+# Now, let us run decoding on the checkpoint generated from our finetune run. Note that the finetune run checkpoint generates the `full state` which has both parameters and optimizer state. For decoding, we only need to use the parameters. So, we can use the `MaxText/generate_param_only_checkpoint.py` to convert
+# the full state checkpoint into a parameter only checkpoint for more efficient memory use. Note that the path provided to the flag `load_full_state_path` is the path to the checkpoint subdirectory inside the `base_output_directory` from our previous finetuning run, say the checkpoint saved at finetuning step #5
+# Also, `force_unroll=true` is converting the output parameter only checkpoint into an unscanned format for efficient decoding
+unset M_LOAD_PARAMETERS_PATH
+export PARAMETER_CHECKPOINT_RUN=generate_param_only_checkpoint_${idx}
+python3 MaxText/generate_param_only_checkpoint.py MaxText/configs/base.yml load_full_state_path=${M_BASE_OUTPUT_DIRECTORY}/runner_finetuning_${idx}/checkpoints/5/items run_name=${PARAMETER_CHECKPOINT_RUN} model_name='llama2-7b' force_unroll=true
+
+# Like before, we define `new_ckpt_path` to refer to the checkpoint subdirectory exactly
+export new_ckpt_path=${M_BASE_OUTPUT_DIRECTORY}/${PARAMETER_CHECKPOINT_RUN}/checkpoints/0/items
+
+# We run decoding on the fine-tuned parameter checkpoint 
+python3 MaxText/decode.py MaxText/configs/base.yml load_parameters_path=${new_ckpt_path} run_name=runner_decode_finetuned_${idx} per_device_batch_size=1 model_name='llama2-7b' ici_autoregressive_parallelism=4 max_prefill_predict_length=4  max_target_length=16 prompt="I love to" attention=dot_product scan_layers=false
+
+# Note that the `converted_checkpoint` is in a `scanned` format which is great for training but for efficient decoding performance we want the checkpoint in an `unscanned` format.
+# We can do this by running `MaxText/generate_param_only_checkpoint.py` on `converted_checkpoint` again with `force_unroll=true`. 
+export DIRECT_PARAMETER_CHECKPOINT_RUN=direct_generate_param_only_checkpoint_${idx}
+python3 MaxText/generate_param_only_checkpoint.py MaxText/configs/base.yml load_parameters_path=${converted_checkpoint} run_name=${DIRECT_PARAMETER_CHECKPOINT_RUN} model_name='llama2-7b' force_unroll=true
+
+# Like before, we define `unscanned_ckpt_path` to refer to the checkpoint subdirectory exactly
```

Comment:
It seems to be like this should go earlier in the file right? Since not fine tuning is kind of a simpler standalone story?

Reply:
And we should say that alternatively you can skip to fine tuning.

Reply:
done